### PR TITLE
Define ParserContext as ref struct

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

File: src/GraphQLParser/ParserContext.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@
 namespace GraphQLParser;
 
 // WARNING: mutable struct, pass it by reference to those methods that will change it
-internal partial struct ParserContext
+internal ref partial struct ParserContext
```
Comment (sungam3r) [inline src/GraphQLParser/ParserContext.cs:9]:
```suggestion
internal partial ref struct ParserContext

```

File: src/GraphQLParser/ParserContext.Parse.cs [hunk 2]
```diff
@@ -4,7 +4,7 @@
 namespace GraphQLParser;
 
 // WARNING: mutable struct, pass it by reference to those methods that will change it
-internal partial struct ParserContext
+internal ref partial struct ParserContext
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQLParser/ParserContext.Parse.cs:7]: oops, CS1585	Member modifier 'ref' must precede the member type and name

File: src/GraphQLParser/ParserContext.Parse.cs [hunk 3]
```diff
@@ -4,7 +4,7 @@
 namespace GraphQLParser;
 
 // WARNING: mutable struct, pass it by reference to those methods that will change it
```
Comment (sungam3r) [inline src/GraphQLParser/ParserContext.Parse.cs:6]:
```suggestion
// WARNING: mutable ref struct, pass it by reference to those methods that will change it
```

File: src/GraphQLParser/ParserContext.cs [hunk 4]
```diff
@@ -6,7 +6,7 @@
 namespace GraphQLParser;
 
 // WARNING: mutable struct, pass it by reference to those methods that will change it
```
Comment (sungam3r) [inline src/GraphQLParser/ParserContext.cs:8]:
```suggestion
// WARNING: mutable ref struct, pass it by reference to those methods that will change it

```

**Review by sungam3r — APPROVED** [review]
🤔 It seemed to me that I once tried to do this. Do you have a practical scenario when it is required? Let's add ref to `LexerContext` too.

**Comment by Shane32** [comment]
> 🤔 It seemed to me that I once tried to do this. Do you have a practical scenario when it is required? Let's add ref to `LexerContext` too.

No. It simply is already a ref struct — it is always (and must be) passed by ref; it’s a struct, and it’s internal so the change does not affect the public api or operation but is simply “good housekeeping”

I didn’t look at LexerContext

**Comment by Shane32** [comment]
The change to ref struct prevents accidental storage on the heap, such as seen by the change to the tests

**Pull request merged**